### PR TITLE
Support dictionary data type in array_to_string

### LINKED ISSUE #10862: `array_to_string` panic on dictionary values
### Describe the bug

`array_to_string` function panic's when running a query on data that has dictionary data type. To fix the panic, `array_to_string` needs to support dictionary type

### To Reproduce

1. Use DataFusion CLI:
```shell
# in dir datafusion/datafusion-cli
$ cargo build
$ ./target/debug/datafusion-cli
```

2. Write sample data
```sql
CREATE TABLE table1 AS VALUES (1, arrow_cast('foo', 'Dictionary(Int32, Utf8)')), (2, arrow_cast('bar', 'Dictionary(Int32, Utf8)'));
```

3. Run query
```sql
> SELECT array_to_string(array_agg(column2),',') FROM (SELECT column2 FROM table1);
thread 'main' panicked at .../datafusion/datafusion/functions-array/src/string.rs:298:17:
internal error: entered unreachable code
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

### Expected behavior

```sql
> SELECT array_to_string(array_agg(column2),',') FROM (SELECT column2 FROM table1);
+------------------------------------------------------+
| array_to_string(ARRAY_AGG(table2.column2),Utf8(",")) |
+------------------------------------------------------+
| foo,bar                                              |
+------------------------------------------------------+
```

### Additional context

1. Here are the data types that `array_to_string` currently supports:
https://github.com/apache/datafusion/blob/59120255916bcd624161ce8f5df255f2cc838406/datafusion/functions-array/src/string.rs#L63-L101

2. instead of panic, it would be better to throw an error

### PR DESCRIPTION
**Pull request by EduardoVega**
## Which issue does this PR close?

Closes #10862 

## Rationale for this change

Go to issue.

## What changes are included in this PR?

Dictionary is added as a supported data type in `compute_array_to_string`, so that it can properly handle a Dictionary Array.

## Are these changes tested?

Yes

### CODE REVIEW

File: datafusion/functions-array/src/string.rs [hunk 1]
```diff
@@ -281,6 +281,21 @@ pub(super) fn array_to_string_inner(args: &[ArrayRef]) -> Result<ArrayRef> {
 
                 Ok(arg)
             }
+            Dictionary(..) => {
+                let any_dict_array = arr.as_any_dictionary_opt().ok_or_else( || {
+                    DataFusionError::Internal(
+                        format!("could not cast {} to AnyDictionaryArray", arr.data_type())
+                    )
+                })?;
+                compute_array_to_string(
```
Comment (alamb) [inline datafusion/functions-array/src/string.rs:290]: I think this code may just return the dictionary converted to string, 

https://docs.rs/arrow/latest/arrow/array/struct.DictionaryArray.html


You might have to expand out the values again (perhaps cast the result to `DataType::Utf8` 🤔

**Review by alamb — COMMENTED** [review]
Thank you su much for this contribution @EduardoVega 

Can you please add some tests, ideally in sqllogicteset format?

Here are the instructions: https://github.com/apache/datafusion/tree/main/datafusion/sqllogictest

Ideally you should be able to extend one of the existing test files in https://github.com/apache/datafusion/tree/main/datafusion/sqllogictest/test_files with the @appletreeisyellow 's reproduer from https://github.com/apache/datafusion/issues/10862

For example, this might be a good spit
https://github.com/apache/datafusion/blob/cc60278f50eac33f9c8ea3509c171b4f2b008b27/datafusion/sqllogictest/test_files/array.slt#L3752

File: datafusion/functions-array/src/string.rs (hunk 1 above)
Comment (EduardoVega) [reply datafusion/functions-array/src/string.rs:290]: Checking

File: datafusion/functions-array/src/string.rs [hunk 2]
```diff
@@ -281,6 +281,31 @@ pub(super) fn array_to_string_inner(args: &[ArrayRef]) -> Result<ArrayRef> {
...
+                            &null_string,
+                            with_null_string,
+                            $ARRAY_TYPE
+                        )
+                    };
+                }
+                call_array_function!(
+                    any_dict_array.values().data_type(),
+                    false
+                )
```
Comment (Weijun-H) [inline datafusion/functions-array/src/string.rs:307]: Why not use
```suggestion
                compute_array_to_string(
                    arg,
                    any_dict_array.values().clone(),
                    delimiter.clone(),
                    null_string.clone(),
                    with_null_string,
                )?;

                Ok(arg)
```

Comment (EduardoVega) [reply datafusion/functions-array/src/string.rs:307]: I wanted to see if I could avoid calling again `compute_array_to_string` and instead compute the result and return inmediately, but calling `compute_array_to_string` is what other data types are doing, so makes sense to follow the same pattern.

File: datafusion/functions-array/src/string.rs [hunk 3]
```diff
@@ -281,6 +281,24 @@ pub(super) fn array_to_string_inner(args: &[ArrayRef]) -> Result<ArrayRef> {
...
             }
+            Dictionary(..) => {
+                let any_dict_array = arr
+                    .as_any_dictionary_opt()
+                    .ok_or_else( || {
+                        DataFusionError::Internal(
+                            format!("could not cast {} to AnyDictionaryArray", arr.data_type())
+                        )
+                    })?;
+                compute_array_to_string(
```
Comment (alamb) [inline datafusion/functions-array/src/string.rs:292]: I pushed some tests showing how this results in incorrect values.

I think you may be able to call `to_string!` directly to iterate over the values of the dictionary.

**Review by alamb — COMMENTED** [review]
Thank you again for this PR @EduardoVega  -- very much appreciated

I still think this PR is not quite correct as it is operating on the dictionary rather than the logical values of the dictionary array. 

I pushed a new commit to your branch with some additional tests that I think show wrong results. 

I can help fix the code too, but figured I would leave the feedback in case you wanted to do so.

File: datafusion/functions-array/src/string.rs (hunk 3 above)
Comment (EduardoVega) [reply datafusion/functions-array/src/string.rs:292]: thanks for the feedback, let me figure it out.

File: datafusion/functions-array/src/string.rs (hunk 1 above)
Comment (EduardoVega) [reply datafusion/functions-array/src/string.rs]: I'll continue the conversation in the comments below

File: datafusion/functions-array/src/string.rs (hunk 3 above)
Comment (EduardoVega) [reply datafusion/functions-array/src/string.rs]: @alamb thanks for the tests, I think I have a better understanding of DictionaryArray. 

Based on my understanding, I converted dictionary keys and values to something iterable. Then using the keys to access the values I got the logical array. I think tests are now showing correct results.

Finally, I am happy to work on any new feedback I receive.

File: datafusion/functions-array/src/string.rs [hunk 4]
```diff
@@ -281,6 +283,49 @@ pub(super) fn array_to_string_inner(args: &[ArrayRef]) -> Result<ArrayRef> {
 
                 Ok(arg)
             }
+            Dictionary(_, _) => {
+                let any_dict_array = arr.as_any_dictionary_opt().ok_or_else(|| {
```
Comment (alamb) [inline datafusion/functions-array/src/string.rs:287]: Thank you @EduardoVega  -- it is very impressive you got this to work

I am somewhat concerned that this is quite a bit of code (it actually handles dictionaries of any type, not just string dictionaries, and that it also supports arbitrary key types -- not just the integer types specified)

I am going to try and simplify this a bit

File: datafusion/functions-array/src/string.rs [hunk 5]
```diff
@@ -281,6 +283,49 @@ pub(super) fn array_to_string_inner(args: &[ArrayRef]) -> Result<ArrayRef> {
...
+                    }};
+                }
+                call_array_function!(any_dict_array.values().data_type(), false);
+
+                let mut keys: Vec<usize> = vec![];
+                macro_rules! array_function {
+                    ($ARRAY_TYPE:ident) => {{
+                        for x in downcast_arg!(any_dict_array.keys(), $ARRAY_TYPE) {
+                            if let Some(x) = x {
+                                keys.push(x.to_string().parse::<usize>().unwrap());
```
Comment (alamb) [inline datafusion/functions-array/src/string.rs:311]: this works, but converts integers to strings just to parse them again -- I think we can just support string dictionaries and avoid this

**Review by alamb — COMMENTED** [review]
Thank you so much for working on this @EduardoVega  -- it is quite a bit more complicated than I realized.

File: datafusion/functions-array/src/string.rs (hunk 4 above)
Comment (alamb) [reply datafusion/functions-array/src/string.rs]: I pushed 5a5397bef that I think simplifies the implementation a bit -- let me know what you think

File: datafusion/functions-array/src/string.rs (hunk 3 above)
Comment (alamb) [reply datafusion/functions-array/src/string.rs:297]: Amazing @EduardoVega  -- thank you so much -- this is quite an impressive first contribution. I didn't realize it was going to be quite as tricky.

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
Thanks again @EduardoVega and @Weijun-H

**Pull request merged**